File: 373. Find K Pairs with Smallest Sums/Solution.py
import heapq

class Solution:
    def kSmallestPairs(self, nums1: List[int], nums2: List[int], k: int) -> List[List[int]]:
        n, m = len(nums1), len(nums2)
        pq = [(nums1[0] + nums2[0], 0, 0)]
        res = []
        seen = set((0, 0))
        while pq and k > 0:
            curr, left, right = heappop(pq)
            res.append([nums1[left], nums2[right]])
            k -= 1
            if left < n - 1 and (left + 1, right) not in seen:
                heappush(pq, (nums1[left + 1] + nums2[right], left + 1, right))
                seen.add((left + 1, right))
            if right < m - 1 and (left, right + 1) not in seen:
                heappush(pq, (nums1[left] + nums2[right + 1], left, right + 1))
                seen.add((left, right + 1))
        return res

File: 373. Find K Pairs with Smallest Sums/Solution.java
class Solution {
    public List<List<Integer>> kSmallestPairs(int[] nums1, int[] nums2, int k) {
        int n = nums1.length, m = nums2.length;
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> a[0] - b[0]);
        Map<Integer, Set<Integer>> seen = new HashMap<>();
        List<List<Integer>> res = new ArrayList<>();
        pq.offer(new int[] { nums1[0] + nums2[0], 0, 0 });
        seen.put(0, new HashSet<>());
        seen.get(0).add(0);
        while (!pq.isEmpty() && k > 0) {
            int[] curr = pq.poll();
            res.add(new ArrayList<>(List.of(nums1[curr[1]], nums2[curr[2]])));
            k--;
            if (curr[1] < n - 1 && (!seen.containsKey(curr[1] + 1) || !seen.get(curr[1] + 1).contains(curr[2]))) {
                pq.offer(new int[] { nums1[curr[1] + 1] + nums2[curr[2]], curr[1] + 1, curr[2]});
                if (!seen.containsKey(curr[1] + 1)) seen.put(curr[1] + 1, new HashSet<>());
                seen.get(curr[1] + 1).add(curr[2]);
            }
            if (curr[2] < m - 1 && (!seen.containsKey(curr[1]) || !seen.get(curr[1]).contains(curr[2] + 1))) {
                pq.offer(new int[] { nums1[curr[1]] + nums2[curr[2] + 1], curr[1], curr[2] + 1});
                if (!seen.containsKey(curr[1])) seen.put(curr[1], new HashSet<>());
                seen.get(curr[1]).add(curr[2] + 1);
            }
        }
        return res;
    }
}
